### PR TITLE
Update and fix examples

### PR DESCRIPTION
**Pull request by alisinabh**
Hey :wave: 

According to [this](https://github.com/elixir-nx/axon/pull/239#issuecomment-1118470841) I have been trying to make the examples for axon to work and this is the resulting PR.

All of the examples seem to be working now. But It would be nice if someone else would test and confirm.

### Test checklist
 - [x] `examples/basics/multi_output_example.exs`
 - [x] `examples/basics/multi_input_example.exs` 
 - [x] `examples/structured/credit_card_fraud.exs`
 - [x] `examples/generative/fashionmnist_autoencoder.exs`
 - [x] `examples/generative/mnist_gan.exs`
 - [ ] `examples/vision/horses_or_humans.exs`
 - [x] `examples/vision/cnn_image_denoising.exs`
 - [x] `examples/vision/resnet50.exs`
 - [x] `examples/vision/cifar10.exs`
 - [x] `examples/vision/mnist.exs`
 - [x] `notebooks/multi_input_example.livemd`
 - [x] `notebooks/mnist.livemd`
 - [x] `notebooks/fashionmnist_autoencoder.livemd`

### CODE REVIEW

**Comment by josevalim** [comment]
@alisinabh EXLA v0.2.1 (not yet released) has a fix for the to_batched_list. We should have a new release soon!

**Comment by alisinabh** [comment]
Thanks @josevalim, I have updated the `Mix.install`s in the examples which use `Nx.to_batched_list` to explicitly install `exla ~> 0.2.1`. I tested one with the current exla main and it works.

**Comment by seanmor5** [comment]
Hi @alisinabh I am going to try to release Axon 0.1 soon, and I was hoping we can include some of the examples from here in the release. Can you update me on any remaining issues/work needed to get this PR merged?

**Comment by alisinabh** [comment]
@seanmor5 Great news about Axon 0.1 :confetti_ball: 

There are still issues here. For example the `vision/mnist.exs` or `vision/cifar10.exs` are still failing because of incompatible tensor backends (#244).

After that is fixed we just need to run the ones that I did not add a check mark to them in the list above and make sure everything is working then we can merge.

**Comment by seanmor5** [comment]
@alisinabh I'm about to merge a PR that will break most of the examples again, but then they should be easy to fix! There will be an additional option added to `Axon.Loop.run` which is an initial state, so if passing keyword options you will also need to provide an initial state:

```
Axon.Loop.run(loop, data, initial_state, opts)
```

For `Axon.Loop.trainer` loops the initial state should be an empty map `%{}`

For `Axon.Loop.evaluator` loops, the initial state should be the model state

**Comment by alisinabh** [comment]
@seanmor5 I have updated the examples in 7d3ec6c and the evaluator loop seems to be working without any issues. :+1: 

However, There is now a tiny dependency issue with the following line that you have modified in the Nx dependency options. Will we remove it from `mix.exs`?

https://github.com/elixir-nx/axon/blob/main/mix.exs#L42
```elixir
...
else
  [github: "elixir-nx/nx", sparse: "nx", override: true, branch: "v0.2"]
end
...
```

**Comment by seanmor5** [comment]
When we release an updated Nx and EXLA it will be removed :)

**Comment by alisinabh** [comment]
Great, I have squashed everything and tested most of the examples. I guess this is ready for review now.

**Comment by seanmor5** [comment]
Hey @alisinabh we recently released the updated EXLA/Nx so there should no longer be any dependency mismatches!

**Comment by alisinabh** [comment]
@seanmor5 Confirmed!
I just had to add names to all the input layers but now everything is working :100:

File: README.md [hunk 1]
```diff
@@ -152,8 +152,8 @@ You'll also likely want to include an `Nx` compiler such as `EXLA` for any pract
 def deps do
   [
     {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon", branch: "main"},
-    {:exla, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "exla", override: true},
-    {:nx, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "nx", override: true}
+    {:exla, "~> 0.2.0"},
+    {:nx, "~> 0.2.0"}
```
Comment (seanmor5) [inline README.md:156]:
```suggestion
    {:exla, "~> 0.2.2"},
    {:nx, "~> 0.2.1"}
```

File: examples/basics/multi_input_example.exs [hunk 2]
```diff
@@ -1,15 +1,16 @@
 Mix.install([
   {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
-  {:exla, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "exla"},
-  {:nx, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "nx", override: true},
+  {:exla, "~> 0.2.0"},
+  {:nx, "~> 0.2.0"}
```
Comment (seanmor5) [inline examples/basics/multi_input_example.exs:4]:
```suggestion
  {:exla, "~> 0.2.2"},
  {:nx, "~> 0.2.1"}
```

File: examples/basics/multi_output_example.exs [hunk 3]
```diff
@@ -1,55 +1,71 @@
...
-# as you are trying to optimize for 2 things at once and optimal solutions for each
-# might not lie on a clean "manifold" - e.g. what's good for one output might not
-# be good for another output
-model
-|> Axon.Loop.trainer([mean_squared_error: 0.5, mean_squared_error: 0.5], :adam)
-|> Axon.Loop.run(data, iterations: 250, epochs: 5)
+Mix.install([
+  {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
+  # {:exla, "~> 0.2.0"},
+  {:nx, "~> 0.2.0"}
```
Comment (seanmor5) [inline examples/basics/multi_output_example.exs:4]:
```suggestion
  # {:exla, "~> 0.2.2"},
  {:nx, "~> 0.2.1"}
```

File: examples/basics/multi_output_example.exs [hunk 4]
```diff
@@ -1,55 +1,71 @@
...
+])
+
+# Specify EXLA as the default defn compiler
+# Nx.Defn.default_options(compiler: EXLA)
```
Comment (seanmor5) [inline examples/basics/multi_output_example.exs:8]:
```suggestion
# EXLA.set_as_nx_default([:tpu, :cuda, :rocm, :host])
```

File: examples/generative/fashionmnist_autoencoder.exs [hunk 5]
```diff
@@ -1,12 +1,12 @@
 Mix.install([
   {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
-  {:exla, github: "elixir-nx/exla", sparse: "exla"},
-  {:nx, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "nx", override: true},
-  {:scidata, "~> 0.1.3"},
+  {:exla, "~> 0.2.1"},
+  {:nx, "~> 0.2.0"},
```
Comment (seanmor5) [inline examples/generative/fashionmnist_autoencoder.exs:4]:
```suggestion
  {:exla, "~> 0.2.2"},
  {:nx, "~> 0.2.1"},
```

File: examples/generative/mnist_gan.exs [hunk 6]
```diff
@@ -1,11 +1,11 @@
 Mix.install([
-  {:axon, github: "elixir-nx/axon"},
-  {:exla, github: "elixir-nx/nx", sparse: "exla"},
-  {:nx, github: "elixir-nx/nx", sparse: "nx", override: true},
-  {:scidata, "~> 0.1.3"}
+  {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
+  {:exla, "~> 0.2.1"},
+  {:nx, "~> 0.2.0"},
+  {:scidata, "~> 0.1.6"}
```
Comment (seanmor5) [inline examples/generative/mnist_gan.exs:5]:
```suggestion
  {:exla, "~> 0.2.2"},
  {:nx, "~> 0.2.1"},
  {:scidata, "~> 0.1.6"}
```

File: examples/structured/credit_card_fraud.exs [hunk 7]
```diff
@@ -1,11 +1,11 @@
 Mix.install([
   {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
-  {:exla, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "exla"},
-  {:nx, path: "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "nx", override: true},
-  {:explorer, "~> 0.1.0-dev", github: "elixir-nx/explorer"}
+  {:exla, "~> 0.2.1"},
+  {:nx, "~> 0.2.0"},
```
Comment (seanmor5) [inline examples/structured/credit_card_fraud.exs:4]:
```suggestion
  {:exla, "~> 0.2.2"},
  {:nx, "~> 0.2.1"},
```

File: examples/vision/cifar10.exs [hunk 8]
```diff
@@ -1,12 +1,12 @@
 Mix.install([
   {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
-  {:exla, github: "elixir-nx/exla", sparse: "exla"},
-  {:nx, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "nx", override: true},
+  {:exla, "~> 0.2.1"},
+  {:nx, "~> 0.2.0", override: true},
```
Comment (seanmor5) [inline examples/vision/cifar10.exs:4]:
```suggestion
  {:exla, "~> 0.2.2"},
  {:nx, "~> 0.2.1"},
```

File: examples/vision/cnn_image_denoising.exs [hunk 9]
```diff
@@ -1,11 +1,11 @@
 Mix.install([
   {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
-  {:exla, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "exla"},
-  {:nx, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "nx", override: true},
-  {:scidata, "~> 0.1.0"}
+  {:exla, "~> 0.2.1"},
+  {:nx, "~> 0.2.0"},
```
Comment (seanmor5) [inline examples/vision/cnn_image_denoising.exs:4]:
```suggestion
  {:exla, "~> 0.2.2"},
  {:nx, "~> 0.2.1"},
```

File: examples/vision/horses_or_humans.exs [hunk 10]
```diff
@@ -1,11 +1,11 @@
 Mix.install([
   {:stb_image, "~> 0.1.0"},
   {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
-  {:exla, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "exla"},
-  {:nx, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "nx", override: true}
+  {:exla, "~> 0.2.0"},
+  {:nx, "~> 0.2.0"}
```
Comment (seanmor5) [inline examples/vision/horses_or_humans.exs:5]:
```suggestion
  {:exla, "~> 0.2.2"},
  {:nx, "~> 0.2.1"}
```

File: examples/vision/mnist.exs [hunk 11]
```diff
@@ -1,18 +1,16 @@
 Mix.install([
   {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
-  {:exla, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "exla"},
-  {:nx, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "nx", override: true},
-  {:scidata, "~> 0.1.3"}
+  {:exla, "~> 0.2.1"},
+  {:nx, "~> 0.2.0"},
```
Comment (seanmor5) [inline examples/vision/mnist.exs:4]:
```suggestion
  {:exla, "~> 0.2.2"},
  {:nx, "~> 0.2.1"},
```

File: notebooks/multi_input_example.livemd [hunk 12]
```diff
@@ -7,8 +7,8 @@ At the very beginning, we have to import essential packages. ```Axon``` will be
 ```elixir
 Mix.install([
   {:axon, "~> 0.1.0-dev", github: "elixir-nx/axon"},
-  {:exla, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "exla"},
-  {:nx, "~> 0.1.0-dev", github: "elixir-nx/nx", sparse: "nx", override: true}
+  {:exla, "~> 0.2.0"},
+  {:nx, "~> 0.2.0"}
```
Comment (seanmor5) [inline notebooks/multi_input_example.livemd:11]:
```suggestion
  {:exla, "~> 0.2.2"},
  {:nx, "~> 0.2.1"}
```

**Review by seanmor5 — APPROVED** [review]
Just the version fixes, newer Axon requires EXLA 0.2.2 :)

File: README.md (hunk 1 above)
Comment (alisinabh) [reply README.md]: Fixed in 64b4365

File: examples/basics/multi_input_example.exs (hunk 2 above)
Comment (alisinabh) [reply examples/basics/multi_input_example.exs]: Fixed in 64b4365

File: examples/basics/multi_output_example.exs (hunk 3 above)
Comment (alisinabh) [reply examples/basics/multi_output_example.exs]: Fixed in 64b4365

File: examples/basics/multi_output_example.exs (hunk 4 above)
Comment (alisinabh) [reply examples/basics/multi_output_example.exs]: Fixed in 64b4365

File: examples/generative/fashionmnist_autoencoder.exs (hunk 5 above)
Comment (alisinabh) [reply examples/generative/fashionmnist_autoencoder.exs]: Fixed in 64b4365

File: examples/generative/mnist_gan.exs (hunk 6 above)
Comment (alisinabh) [reply examples/generative/mnist_gan.exs]: Fixed in 64b4365

File: examples/structured/credit_card_fraud.exs (hunk 7 above)
Comment (alisinabh) [reply examples/structured/credit_card_fraud.exs]: Fixed in 64b4365

File: examples/vision/cifar10.exs (hunk 8 above)
Comment (alisinabh) [reply examples/vision/cifar10.exs]: Fixed in 64b4365

File: examples/vision/cnn_image_denoising.exs (hunk 9 above)
Comment (alisinabh) [reply examples/vision/cnn_image_denoising.exs]: Fixed in 64b4365

File: examples/vision/horses_or_humans.exs (hunk 10 above)
Comment (alisinabh) [reply examples/vision/horses_or_humans.exs]: Fixed in 64b4365

File: examples/vision/mnist.exs (hunk 11 above)
Comment (alisinabh) [reply examples/vision/mnist.exs]: Fixed in 64b4365

File: notebooks/multi_input_example.livemd (hunk 12 above)
Comment (alisinabh) [reply notebooks/multi_input_example.livemd]: Fixed in 64b4365

**Pull request merged**